### ISSUE TITLE
flake: TestGarbageCollectorTypesFromBinding

### ISSUE THREAD

**Issue by hardys** [kcp-dev/kcp#2665]
### Describe the bug

Seeing [failures](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/kcp-dev_kcp/2659/pull-ci-kcp-dev-kcp-main-e2e-shared/1617467644934361088) like: 

```
{Failed  === RUN   TestGarbageCollectorTypesFromBinding/tc1
=== PAUSE TestGarbageCollectorTypesFromBinding/tc1
=== CONT  TestGarbageCollectorTypesFromBinding/tc1
=== CONT  TestGarbageCollectorTypesFromBinding/tc1
    garbagecollector_test.go:153: Created root:universal workspace root:e2e-workspace-xm267:gc-api-binding-1 as /clusters/2hgmlhj8joun35pp
    garbagecollector_test.go:155: Create a binding in the user workspace
    garbagecollector_test.go:179: Wait for the binding to be ready
=== CONT  TestGarbageCollectorTypesFromBinding/tc1
    garbagecollector_test.go:196: Wait for being able to list cowboys in the user workspace
=== CONT  TestGarbageCollectorTypesFromBinding/tc1
    garbagecollector_test.go:206: Creating owner cowboy
=== CONT  TestGarbageCollectorTypesFromBinding/tc1
    garbagecollector_test.go:217: Creating owned configmap
=== CONT  TestGarbageCollectorTypesFromBinding/tc1
    garbagecollector_test.go:228: Creating owned cowboy
    garbagecollector_test.go:247: Deleting owner cowboy
=== CONT  TestGarbageCollectorTypesFromBinding/tc1
    garbagecollector_test.go:251: Waiting for the owned configmap to be garbage collected
=== CONT  TestGarbageCollectorTypesFromBinding/tc1
    assertions.go:1691: Waiting for condition, but got: configmap not garbage collected: owned
=== CONT  TestGarbageCollectorTypesFromBinding/tc1
    garbagecollector_test.go:252: 
        	Error Trace:	util.go:318
        	            				garbagecollector_test.go:252
        	Error:      	Condition never satisfied
        	Test:       	TestGarbageCollectorTypesFromBinding/tc1
        	Messages:   	error waiting for owned configmap to be garbage collected
    --- FAIL: TestGarbageCollectorTypesFromBinding/tc1 (39.68s)
```

`TestGarbageCollectorNormalCRDs` also fails in a similar way

### Steps To Reproduce

Flake observed in CI on https://github.com/kcp-dev/kcp/pull/2659 - can't reproduce the issue locally

### Expected Behaviour

Need to observe CI flake rate and see if this is a one-off or warrants deeper investigation

**Comment by ncdc** [comment]
Looking into this with @nrb - we see kcp's garbage collector controller queues the key for the logical cluster in question multiple times, but we never see the `processing key` log message from the worker goroutine (from `processNextWorkItem`) for the logical cluster in question. We think there is either a deadlock somewhere, or the workers are just so overwhelmed they can't clear the backlog within `wait.ForeverTestTimeout`.

**Comment by ncdc** [comment]
I believe this was fixed by #2751
